D/D.sublime-syntax

```diff
@@ -8,339 +8,1970 @@ file_extensions:
+  type_qualifier_lookahead: '(?=\b({{type_qualifier}})\b)'
+  attribute_lookahead: '(?=@|\b({{attributes}}|extern|align|deprecated|pragma|private|protected|public|export|package)\b)'
+  attributes: 'static|abstract|final|override|synchronized|scope|__gshared|nothrow|pure|ref|return|auto'
+  at_attributes: 'disable|nogc|property|safe|system|trusted'
+  parameter_attribute_lookahead: '(?=\b({{parameter_attribute}})\b)'
+  parameter_attribute: 'in|lazy|out|alias'
+
+  construction_lookahead: '(?=~?\s*this\s*\()'
+
+  keyword: 'else|enum|export|extern|debug|default|delegate|delete|deprecated|do|body|break|case|cast|catch|class|const|continue|abstract|alias|align|asm|assert|auto|final|finally|for|foreach|foreach_reverse|function|goto|if|immutable|import|in|inout|interface|invariant|is|lazy|macro|mixin|module|new|nothrow|out|override|package|pragma|private|protected|public|pure|ref|return|scope|shared|static|struct|switch|synchronized|template|throw|try|typeid|typeof|union|unittest|version|while|with|__gshared|__traits|__vector|__parameters'
+  basic_type: 'bool|byte|cdouble|cent|cfloat|char|creal|dchar|double|float|idouble|ifloat|int|ireal|long|real|short|ubyte|ucent|uint|ulong|ushort|void|wchar|string|dstring|wstring'
+  language_constant: 'null|true|false|__FILE__|__FILE_FULL_PATH__|__MODULE__|__LINE__|__FUNCTION__|__PRETTY_FUNCTION__|__DATE__|__EOF__|__TIME__|__TIMESTAMP__|__VENDOR__|__VERSION__|__ctfe'
+  language_variable: 'this|super'
+  reserved: '{{keyword}}|{{basic_type}}|{{language_constant}}|{{language_variable}}'
+
+  operator_overloads: 'opNeg|opCom|opPostInc|opPostDec|opCast|opAdd|opSub|opSub_r|opMul|opDiv|opDiv_r|opMod|opMod_r|opAnd|opOr|opXor|opShl|opShl_r|opShr|opShr_r|opUShr|opUShr_r|opCat|opCat_r|opEquals|opEquals|opCmp|opCmp|opCmp|opCmp|opAddAssign|opSubAssign|opMulAssign|opDivAssign|opModAssign|opAndAssign|opOrAssign|opXorAssign|opShlAssign|opShrAssign|opUShrAssign|opCatAssign|opIndex|opIndexAssign|opCall|opSlice|opSliceAssign|opPos|opAdd_r|opMul_r|opAnd_r|opOr_r|opXor_r'
+
+  type_specialization_keyword: '__vector|function|delegate|super|return|__parameters'
+  type_specialization_storage: 'struct|union|class|interface|enum'
+
+  block_statement_loohahead: '(?={)'
+
 contexts:
+  prototype:
+    - include: comment-in
+
   main:
-    - match: /\*\*/
-      scope: comment.block.empty.d punctuation.definition.comment.d
-    - include: scope:text.html.javadoc
-    - match: |-
-        (?x)^\s*
-        ((?:\b(private|package|protected|public|export|static|final|native|synchronized|abstract)\b\s*)*) # modifier
-        (class|interface)\s+
-        (\w+)\s* # identifier
-        (?:\(\s*([^\)]+)\s*\)|)\s* # Template type
-        (?:
-          \s*(:)\s*
-          (\w+)
-          (?:\s*,\s*(\w+))?
-          (?:\s*,\s*(\w+))?
-          (?:\s*,\s*(\w+))?
-          (?:\s*,\s*(\w+))?
-          (?:\s*,\s*(\w+))?
-          (?:\s*,\s*(\w+))?
-        )? # super class
+    - meta_include_prototype: false
+    - match: '^#!.*$'
```

Comment:
`#` should be scoped as `punctuation.definition.comment.number-sign` to match the current shell syntax.

Otherwise I am not sure about the global strategy to support shebang directly.

The main shell scope is `source.shell` The bash syntax currently uses `soure.shell.bash`.

To follow this kind of scheme of different shell types any syntax would need to use `source.shell.<syntax>`.

Example: `source.shell.python`, `source.shell.erlang`, `source.shell.tcl`, ... .

Therefore each syntax, which wants to assign itself as a shell type, should provide a dedicated `<Syntax> Shell.sublime-syntax`, I think. (Btw. might be worth a RFC discussion?).

@all: Maybe something like that would be worth thinking about?

```YAML
%YAML 1.2
---
# http://www.sublimetext.com/docs/3/syntax.html
name: D Shell (D)
file_extensions:
  - ???
first_line_match: |-
  (?x:
    ^ \s* \#! .* \bD\b |            # shebang
    ^ \s* \#+ \s* -\*- .*? \bD\b .*? -\*-  # editorconfig
  )
scope: source.shell.d

contexts:

  main:
    - match: ^\s*(\#)
      captures:
        1:
          comment.line.number-sign.shell
          punctuation.definition.comment.shell
      push:
        - meta_include_prototype: false
        - meta_content_scope: comment.line.number-sign.shell
        - match: \n
          scope: comment.line.number-sign.shell
          pop: true
    - match: ^
      set: scope:source.d

  prototype:
    - include: scope:source.d#prototype
```

_Note: We propably should not be too restrictive with whitespace in front of shebangs?_

Reply:
In relation to the whitespace restriction in front of the `#!`: putting a space in front fails to compile with D.

Reply:
I'm not sure why we'd want a separate syntax for these files. They are still just files in D with a shebang. Imo just highlight the shebang line as a comment, assign appropriate scopes and be done with it.

We could probably unify this across syntaxes as `comment.line.number-sign.shebang` and special-case it.

Reply:
I agree with @FichteFoll, KISS!

Reply:
I am fine with this strategy as well.

But anyway: the current solution breaks `embed` statements, by calling `set: ...` to leave the `main` context.

The `escape` of the following statement would be ignored.

```YAML
    - match: ''
      embed: scope:source.d
      embed_scope: source.d.embedded.markdown
      escape: (?=^\`{3})
```

_Note: Found it while applying that strategy to Erlang._

The JavaScript syntax uses `push` which works fine with `embed`.

So something like that might work

```YAML
contexts:

  main:
    - match: ''
      push: [main-after-shebang, comment-shebang]

  comment-shebang:
    - match: ^\#!
      scope: punctuation.definition.comment.shell
      set:
        - meta_include_prototype: false
        - meta_scope: comment.line.number-sign.shebang.d
        - match: \bD\b
          scope: constant.language.d
        - match: \n
          pop: true

  main-after-shebang:
```

Reply:
Pushing away might on the other hand cause issues, if the syntax is `include`ed into another one, as the `main-after-shebang` is never popped off.

With that said: This kind of shebang handling is problematic for syntaxes which might want to be ebedded/included into other ones.

Reply:
see: https://github.com/SublimeTextIssues/Core/issues/2395